services/leader/client.go

```diff
@@ -0,0 +1,249 @@
+	// import etcd's concurrency package in order to check the cause of the
+	// error.
+	ErrNoLeader = concurrency.ErrElectionNoLeader
+
+	// ErrCampaignInProgress is returned when a call to Campaign() is made while
+	// the caller is either already (a) campaigning or (b) the leader.
+	ErrCampaignInProgress = errors.New("campaign in progress")
+)
+
+// NB(mschalle): when an etcd leader failover occurs, all current leases have
+// their TTLs refreshed: https://github.com/coreos/etcd/issues/2660
+
+type client struct {
+	sync.RWMutex
+
+	client      *election.Client
+	opts        services.ElectionOptions
+	cancelFn    context.CancelFunc
+	resignCh    chan struct{}
+	campaigning bool
+	closed      bool
+}
+
+// newClient returns an instance of an client bound to a single election.
+func newClient(cli *clientv3.Client, opts Options, electionID string) (*client, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	ttl := opts.ElectionOpts().TTL()
+	pfx := electionPrefix(opts.ServiceID(), electionID)
+	ec, err := election.NewClient(cli, pfx, election.WithSessionOptions(concurrency.WithTTL(ttl)))
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		client:   ec,
+		opts:     opts.ElectionOpts(),
+		resignCh: make(chan struct{}),
+	}, nil
+}
+
+func (c *client) campaign(opts services.CampaignOptions) (<-chan campaign.Status, error) {
+	if c.isClosed() {
+		return nil, errClientClosed
+	}
+
+	if !c.startCampaign() {
+		return nil, ErrCampaignInProgress
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	c.Lock()
+	c.cancelFn = cancel
+	c.Unlock()
+
+	proposeVal := c.val(opts.LeaderValue())
+
+	// buffer 1 to not block initial follower update
+	sc := make(chan campaign.Status, 1)
+
+	sc <- campaign.NewStatus(campaign.Follower)
+
+	go func() {
+		defer func() {
+			close(sc)
+			cancel()
+			c.resetCampaign()
+		}()
+
+		// Campaign blocks until elected. Once we are elected, we get a channel
+		// that's closed if our session dies.
+		ch, err := c.client.Campaign(ctx, proposeVal)
+		if err != nil {
+			sc <- campaign.NewErrorStatus(err)
+			return
+		}
+
+		sc <- campaign.NewStatus(campaign.Leader)
+		select {
+		case <-ch:
+			sc <- campaign.NewErrorStatus(election.ErrSessionExpired)
+		case <-c.resignCh:
+			sc <- campaign.NewStatus(campaign.Follower)
+		}
+	}()
+
+	return sc, nil
+}
+
+func (c *client) resign() error {
+	if c.isClosed() {
+		return errClientClosed
+	}
+
+	// if there's an active blocking call to Campaign() stop it
+	c.Lock()
+	if c.cancelFn != nil {
+		c.cancelFn()
+		c.cancelFn = nil
+	}
+	c.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.opts.ResignTimeout())
+	defer cancel()
+	if err := c.client.Resign(ctx); err != nil {
+		return err
+	}
+
+	// if successfully resigned and there was a campaign in Leader state cancel
+	// it
+	select {
+	case c.resignCh <- struct{}{}:
+	default:
+	}
+
+	c.resetCampaign()
```

Comment:
does resetCampaign need to be called in both here and in the go routine in campaign()?